### PR TITLE
fix(tiering): Throttle snapshot load

### PR DESCRIPTION
**Pull request by dranikpg**

### CODE REVIEW

File: src/server/tiered_storage.cc [hunk 1]
```diff
@@ -377,11 +383,9 @@ void TieredStorage::RunOffloading(DbIndex dbid) {
   PrimeTable::Cursor start_cursor{};
 
   // Loop while we haven't traversed all entries or reached our stash io device limit.
-  // Keep number of iterations below resonable limit to keep datastore always responsive
-  size_t iterations = 0;
   do {
     offloading_cursor_ = table.TraverseBySegmentOrder(offloading_cursor_, cb);
-  } while (offloading_cursor_ != start_cursor && stash_limit > 0 && iterations++ < 100);
+  } while (offloading_cursor_ != start_cursor && stash_limit > 0);
```
Comment (dranikpg) [inline src/server/tiered_storage.cc:388]: the dilemma is that 100 is not enough to keep the disk busy during the loading phase. My options are finding some reasonable settings for the general case, or calling RunOffloading() from the rdb loader to keep it busy.

**Comment by dranikpg** [comment]
@romange  PTAL only at the design, will polish if this approach is OK

File: src/server/tiered_storage.cc (hunk 1 above)
Comment (romange) [reply src/server/tiered_storage.cc:388]: what about calling Stash directly like you do in SET ?

Comment (dranikpg) [reply src/server/tiered_storage.cc:388]: Added, the problem is that when we issue too many stashes, it overflows the limit, so we still rely on background offloading to finish the rest 🤔

File: src/server/rdb_load.cc [hunk 2]
```diff
@@ -2466,6 +2478,9 @@ void RdbLoader::LoadItemsBuffer(DbIndex db_ind, const ItemsBuf& ib) {
     if (!res.is_new) {
       LOG(WARNING) << "RDB has duplicated key '" << item->key << "' in DB " << db_ind;
     }
+
+    if (auto* ts = es->tiered_storage(); ts && ts->ShouldStash(res.it->second))
+      ts->Stash(db_cntx.db_index, item->key, &res.it->second);
```
Comment (romange) [inline src/server/rdb_load.cc:2483]: nit: Consider merging `ShouldStash` and `Stash` into `TryStash` or something

File: src/server/rdb_load.h [hunk 3]
```diff
@@ -270,7 +270,7 @@ class RdbLoader : protected RdbLoaderBase {
  private:
   Service* service_;
   ScriptMgr* script_mgr_;
-  std::unique_ptr<ItemsBuf[]> shard_buf_;
+  std::vector<ItemsBuf> shard_buf_;
```
Comment (romange) [inline src/server/rdb_load.h:273]: curious, why the change?

**Comment by romange** [comment]
Did you try loading a dataset with X GB into a datastore with tiering and maxmemory Y, where Y <<X?
If yes, can you share some numbers?

**Comment by dranikpg** [comment]
> Did you try loading a dataset with X GB into a datastore with tiering and maxmemory Y, where Y <<X?
If yes, can you share some numbers?

I have a skipped test in the PR. You can run it in any scenario, just alter debug populate and optionally add a tiering limit. Depends heavily on the write depth. Also see my comment that's it's still easily breakable

**Comment by romange** [comment]
lgtm but it could be nice if you could run it one or two configs yourself and provide an anecdotal info on how well it works.

File: src/server/rdb_load.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/rdb_load.cc]: Renamed

File: src/server/rdb_load.h (hunk 3 above)
Comment (dranikpg) [reply src/server/rdb_load.h:273]: Just to keep the code less confusing 🙂

File: src/server/tiered_storage.cc [hunk 4]
```diff
@@ -271,14 +271,15 @@ template util::fb2::Future<size_t> TieredStorage::Modify(DbIndex dbid, std::stri
                                                          const PrimeValue& value,
                                                          std::function<size_t(std::string*)> modf);
 
-void TieredStorage::Stash(DbIndex dbid, string_view key, PrimeValue* value) {
-  CHECK(!value->IsExternal() && !value->HasIoPending());
+bool TieredStorage::TryStash(DbIndex dbid, string_view key, PrimeValue* value) {
+  if (!ShouldStash(*value))
+    return false;
```
Comment (romange) [inline src/server/tiered_storage.cc:276]: please copy the `CHECK`

**Review by romange — COMMENTED** [review]
lgtm

File: src/server/tiered_storage.cc (hunk 4 above)
Comment (dranikpg) [reply src/server/tiered_storage.cc:276]: ShouldStash checks all conditions of the CHECK

**Review by romange — APPROVED** [review]
